src/niri.rs

```diff
@@ -1209,12 +1214,22 @@ impl Niri {
                     if let Some((surface, bbox)) = dnd {
                         if let Some(mut overlap) = geo.intersection(bbox) {
                             overlap.loc -= surface_pos;
+                            dnd_scale = cursor_scale.max(output.current_scale().integer_scale());
                             output_update(output, Some(overlap), surface);
                         } else {
                             output_update(output, None, surface);
                         }
                     }
                 }
+
+                if surface.version() >= 6 {
+                    surface.preferred_buffer_scale(cursor_scale);
+                }
+                if let Some((surface, _)) = dnd {
+                    if surface.version() >= 6 {
+                        surface.preferred_buffer_scale(dnd_scale);
```

Comment:
There's also dnd surface handling in the branch above, also needs to be handled.

Reply:
Is there some other `dnd` surface other the one in `dnd` struct?

Reply:
I mean the `CursorImageStatus::Hidden | CursorImageStatus::Named(_)` branch above, which may still have a DnD surface. It does not get the preferred buffer scale yet as far as I can see.